src/network/network_coordinator.cpp

```diff
@@ -0,0 +1,227 @@
+}
+
+void ClientNetworkCoordinatorSocketHandler::Connect()
+{
+	/* We are either already connected or are trying to connect. */
+	if (this->sock != INVALID_SOCKET || this->connecting) return;
+
+	this->Reopen();
+
+	this->connecting = true;
+	new NetworkCoordinatorConnecter(NETWORK_COORDINATOR_SERVER_HOST);
+}
+
+NetworkRecvStatus ClientNetworkCoordinatorSocketHandler::CloseConnection(bool error)
+{
+	NetworkCoordinatorSocketHandler::CloseConnection(error);
+
+	this->CloseSocket();
+	this->connecting = false;
+
+	_network_server_connection_type = CONNECTION_TYPE_UNKNOWN;
+	this->next_update = {};
+
+	SetWindowDirty(WC_CLIENT_LIST, 0);
+
+	return NETWORK_RECV_STATUS_OKAY;
+}
+
+/**
+ * Register our server to receive our join-key.
+ */
+void ClientNetworkCoordinatorSocketHandler::Register()
+{
+	_network_server_connection_type = CONNECTION_TYPE_UNKNOWN;
+	this->next_update = {};
+
+	SetWindowDirty(WC_CLIENT_LIST, 0);
+
+	this->Connect();
+
+	Packet *p = new Packet(PACKET_COORDINATOR_CLIENT_REGISTER);
+	p->Send_uint8(NETWORK_COORDINATOR_VERSION);
+	p->Send_uint8(SERVER_GAME_TYPE_PUBLIC);
+	p->Send_uint16(_settings_client.network.server_port);
+
+	this->SendPacket(p);
+}
+
+/**
+ * Send an update of our server status to the Game Coordinator.
+ */
+void ClientNetworkCoordinatorSocketHandler::SendServerUpdate()
+{
+	Debug(net, 6, "Sending server update to Game Coordinator");
+	this->next_update = std::chrono::steady_clock::now() + NETWORK_COORDINATOR_DELAY_BETWEEN_UPDATES;
+
+	Packet *p = new Packet(PACKET_COORDINATOR_CLIENT_UPDATE);
+	p->Send_uint8(NETWORK_COORDINATOR_VERSION);
+	SerializeNetworkGameInfo(p, GetCurrentNetworkServerGameInfo());
+
+	this->SendPacket(p);
+}
+
+/**
+ * Check whether we received/can send some data from/to the Game Coordinator server and
+ * when that's the case handle it appropriately
+ */
+void ClientNetworkCoordinatorSocketHandler::SendReceive()
+{
+	/* Private games are not listed via the Game Coordinator. */
+	if (_network_server && !_settings_client.network.server_advertise) {
+		if (this->sock != INVALID_SOCKET) {
+			this->CloseConnection();
+		}
+		return;
+	}
+
+	static int last_attempt_backoff = 1;
+
+	if (this->sock == INVALID_SOCKET) {
+		static std::chrono::steady_clock::time_point last_attempt = {};
+
+		/* Don't auto-reconnect when we are not a server. */
+		if (!_network_server) return;
+		/* Don't reconnect if we are connecting. */
+		if (this->connecting) return;
+		/* Throttle how often we try to reconnect. Otherwise after an outage
+			* all servers reconnect at the same time, potentially overwhelming
+			* the Game Coordinator. */
+		if (std::chrono::steady_clock::now() < last_attempt + std::chrono::seconds(1) * last_attempt_backoff) return;
+
+		last_attempt = std::chrono::steady_clock::now();
+		/* Delay reconnecting with up to 32 seconds. */
+		if (last_attempt_backoff < 32) {
+			last_attempt_backoff *= 2;
+		}
+
+		Debug(net, 1, "Connection with Game Coordinator lost; reconnecting ...");
```

Comment:
Excluding lang all (55) but two cases of "..." in/near strings did not have a space in front. For consistency that space should not be here either I guess.
`grep -R '\.\.\.' .|grep '"'|grep -v /lang/`